Description: cherry-pick "reader: make PY_SSIZE_T_CLEAN"
Author:      Zbigniew Jędrzejewski-Szmek
Bug-Ubuntu:  https://bugs.launchpad.net/bugs/1963582
Origin:      https://github.com/systemd/python-systemd/commit/c71bbac357f0ac722e1bcb2edfa925b68cca23c9
Forwarded:   not-needed
Last-Update: 2022-03-03
--- a/systemd/_reader.c
+++ b/systemd/_reader.c
@@ -18,7 +18,12 @@
   along with python-systemd; If not, see <http://www.gnu.org/licenses/>.
 ***/
 
+#define PY_SSIZE_T_CLEAN
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wredundant-decls"
 #include <Python.h>
+#pragma GCC diagnostic pop
+
 #include <structmember.h>
 #include <datetime.h>
 #include <time.h>
@@ -707,11 +712,17 @@
              "Match is a string of the form \"FIELD=value\".");
 static PyObject* Reader_add_match(Reader *self, PyObject *args, PyObject *keywds) {
         char *match;
-        int match_len, r;
+        Py_ssize_t match_len;
+        int r;
         if (!PyArg_ParseTuple(args, "s#:add_match", &match, &match_len))
                 return NULL;
 
-        r = sd_journal_add_match(self->j, match, match_len);
+        if (match_len > INT_MAX) {
+                set_error(-ENOBUFS, NULL, NULL);
+                return NULL;
+        }
+
+        r = sd_journal_add_match(self->j, match, (int) match_len);
         if (set_error(r, NULL, "Invalid match") < 0)
                 return NULL;
 
